pkg/solana/txm/pendingtx.go

```diff
@@ -371,17 +377,60 @@ func (c *pendingTxContext) OnFinalized(sig solana.Signature, retentionTimeout ti
 		if retentionTimeout == 0 {
 			return id, nil
 		}
-		// set the timestamp till which the tx should be retained in storage
-		tx.retentionTs = time.Now().Add(retentionTimeout)
-		// update tx state to Finalized
-		tx.state = Finalized
+		finalizedTx := finishedTx{
+			id:          id,
+			state:       Finalized,
+			retentionTs: time.Now().Add(retentionTimeout),
+		}
 		// move transaction from confirmed to finalized map
-		c.finalizedErroredTxs[id] = tx
+		c.finalizedErroredTxs[id] = finalizedTx
+		return id, nil
+	})
+}
+
+func (c *pendingTxContext) OnPrebroadcastError(id string, retentionTimeout time.Duration, txState TxState, _ TxErrType) error {
+	// nothing to do if retention timeout is 0 since transaction is not stored yet.
+	if retentionTimeout == 0 {
+		return nil
+	}
+	err := c.withReadLock(func() error {
+		if tx, exists := c.finalizedErroredTxs[id]; exists && tx.state == txState {
+			return ErrAlreadyInExpectedState
```

Comment:
Better to return an error than idempotency here?

Reply:
I think returning nil here and below where we have the same check wouldn't be idempotent. If we do, we would continue into the write portion and overwrite the existing finishedTx with a renewed retention time. Maybe you could make the argument that we should renew it because a transaction was retried with the same id. But if there's a stored status already, the tx probably shouldn't have been retried. So think it's better to avoid renewing the time in case there's some erroneous retry logic that keeps these errored transactions alive indefinitely. Although this does help me realize I should separate the id exists check and the txState check here.